### PR TITLE
Fix Counter, Leaderboard and TimeSeries bugs

### PR DESCRIPTION
**Pull request by felipecadavid**
This PR fixes some bugs found while testing the components in the example apps.

## TimeSeries
- Connected mode re-fetched data on styles change
- Connected mode re-fetching on every re-render due to useEffect object dependencies

## Leaderboard
- Connected mode re-fetched data on styles change
- Connected mode re-fetching on every re-render due to useEffect object dependencies
- Switching from variant `bar` to `table` and then to `bar` again made `bar` chart disappear

## Counter
- Connected mode re-fetched data on styles change
- Connected mode re-fetching on every re-render due to useEffect object dependencies
- When Connected mode receiving a `null` value, it showed an error, changed to `-` instead
- Sometimes Counter attempted to fetch data on Static mode, added !isStatic condition to fix

### CODE REVIEW

File: packages/react/counter/src/Counter.tsx [hunk 1]
```diff
@@ -118,18 +136,19 @@ export function Counter(props: CounterProps) {
...
+        const fetchedValue = await fetchData()
 
-      if (!fetchedValue) {
-        setHasError(true)
-        console.error(`QueryError: Your metric ${query?.metric} returned undefined or a \`null\` value.`)
-        return
-      }
+        if (typeof fetchedValue === 'undefined') {
+          setHasError(true)
+          console.error(`QueryError: Your metric ${query?.metric} returned undefined.`)
```
Comment (jonatassales) [inline packages/react/counter/src/Counter.tsx:146]: @markandrus I think we should put in discussion a better way to structure error messages. @markandrus do you have any suggestion we could do on a separate PR?

File: packages/react/leaderboard/src/Leaderboard.tsx [hunk 2]
```diff
@@ -196,18 +199,26 @@ export function Leaderboard(props: LeaderboardProps) {
     try {
       setIsLoading(true)
 
+      const dimensions = JSON.parse(dimensionsString)
+      const filters = JSON.parse(filtersString)
```
Comment (jonatassales) [inline packages/react/leaderboard/src/Leaderboard.tsx:203]: We're using `JSON.stringfy` here and then `JSON.parse`. Why?

File: packages/react/time-series/src/TimeSeries.tsx [hunk 3]
```diff
@@ -257,15 +259,22 @@ export function TimeSeries(props: TimeSeriesProps) {
     try {
       setIsLoading(true)
 
+      const filters = JSON.parse(filtersString)
```
Comment (jonatassales) [inline packages/react/time-series/src/TimeSeries.tsx:262]: same transformation here. Maybe I'm missing something?

```suggestion
      const filters = query?.filters || []
```

**Review by jonatassales — CHANGES_REQUESTED** [review]
Looking good. I think we could improve some parts

File: packages/react/counter/src/Counter.tsx [hunk 4]
```diff
@@ -118,18 +136,19 @@ export function Counter(props: CounterProps) {
...
-      const fetchedValue = await fetchData()
+      if (!isStatic) {
+        const fetchedValue = await fetchData()
 
-      if (!fetchedValue) {
-        setHasError(true)
-        console.error(`QueryError: Your metric ${query?.metric} returned undefined or a \`null\` value.`)
-        return
-      }
+        if (typeof fetchedValue === 'undefined') {
```
[the remark this answers is not in the document]
Comment (felipecadavid) [reply packages/react/counter/src/Counter.tsx:144]: This is related to an issue we found in the sample app.
The Counter can return `0` and that would be a normal case, but since `0` is a falsy value it would incorrectly throw this error

File: packages/react/time-series/src/TimeSeries.tsx (hunk 3 above)
Comment (felipecadavid) [reply packages/react/time-series/src/TimeSeries.tsx:262]: The reason for this is that we cannot pass `filters` directly in the dependency array, since `filters` is an array that is not necessarily memoized, it would cause the `useEffect` to trigger on every re-render.

A solution for this is transforming the array to a string, which can be compared  using `===`.

Another solution would be making a deep comparison using a library like [use-deep-compare-effect](https://github.com/kentcdodds/use-deep-compare-effect) but we'd need to add a new dependency and deep comparisons may end up being very costly.

https://github.com/facebook/react/issues/14476#issuecomment-471199055

File: packages/react/counter/src/Counter.tsx [hunk 5]
```diff
@@ -64,14 +66,21 @@ export function Counter(props: CounterProps) {
...
         {
           uniqueName: query?.metric,
           counterInput: {
-            timeRange: query?.timeRange,
-            filters: query?.filters,
+            timeRange: {
+              relative: query?.timeRange?.relative || null,
+              n: query?.timeRange?.n || null,
+              start: query?.timeRange?.start || null,
+              stop: query?.timeRange?.stop || null
```
Comment (markandrus) [inline packages/react/counter/src/Counter.tsx:81]: It might be nice to start using `??` here.

File: packages/react/counter/src/Counter.tsx (hunk 4 above)
Comment (markandrus) [inline packages/react/counter/src/Counter.tsx:144]: @felipecadavid how about this instead?
```suggestion
        if (fetchedValue === undefined) {
```

File: packages/react/counter/src/Counter.tsx (hunk 1 above)
Comment (markandrus) [inline packages/react/counter/src/Counter.tsx:146]: @felipecadavid I am think we should probably remove the `console.error` for now, and plan for logging as a feature soon.

What I mean is that,

1. First of all, customers should be able to enable/disable logging; and
2. Second of all, customers may even want to provide their own callbacks to avoid sending directly to console.

Since we don't have that prepared yet, let's not add more non-configurable `console.error`, `console.log`, etc., calls to the library just yet.
```suggestion
```

File: packages/react/counter/src/utils.ts [hunk 6]
```diff
@@ -25,7 +29,7 @@ export const getValueWithPrefixAndSufix = (params: {
 }) => {
   const { prefix, value, sufix, localize } = params
 
-  if (!value) return
+  if (typeof value === 'undefined') return
```
Comment (markandrus) [inline packages/react/counter/src/utils.ts:32]: How about
```suggestion
  if (value === undefined) return
```

**Review by markandrus — COMMENTED** [review]
@felipecadavid +1 for the detailed PR description. Let's chat tomorrow about

1. How we document feature additions/bug fixes in this project.
    - We might want a pull request template.
    - We might start tracking a CHANGELOG.md in this project.
2. How we manually validate (with @davepropel) that a PR like this is good-to-go.

File: packages/react/counter/src/Counter.tsx (hunk 5 above)
Comment (felipecadavid) [reply packages/react/counter/src/Counter.tsx:81]: Yes, I think that makes sense here

File: packages/react/counter/src/Counter.tsx (hunk 4 above)
Comment (felipecadavid) [reply packages/react/counter/src/Counter.tsx]: Yes, that should work the same way, I'm refactoring this

File: packages/react/counter/src/Counter.tsx (hunk 1 above)
Comment (felipecadavid) [reply packages/react/counter/src/Counter.tsx:146]: I'm going to comment all the `console.error` and will leave a comment related to this

File: packages/react/leaderboard/src/Leaderboard.tsx [hunk 7]
```diff
@@ -317,6 +321,12 @@ export function Leaderboard(props: LeaderboardProps) {
     }
   }, [])
 
+  React.useEffect(() => {
+    if (variant === 'table') {
+      destroyChart()
+    }
```
[the remark this answers is not in the document]
Comment (felipecadavid) [reply packages/react/leaderboard/src/Leaderboard.tsx:327]: We currently only have `'table'` and `'bar'` variants for leaderboard, what I'm doing here is stating that when the `variant` changes to `'table'` we should then cleanup the `'bar'` chart, so we allow changing the component's variant dynamically

**Review by markandrus — APPROVED** [review]
+1 — let's get this in a release candidate
